### PR TITLE
sys: random: fortuna csprng

### PR DESCRIPTION
**Pull request by basilfx**
This PR adds support for the [Fortuna Computationally Secure PRNG](https://www.schneier.com/academic/paperfiles/fortuna.pdf). I had been working on this because of [this application note](https://www.silabs.com/Support%20Documents/TechnicalDocs/AN0806.pdf). I have based this on [this code](https://github.com/blin00/os/blob/master/src/random.c), but fully refactored it for RIOT-OS, while comparing it to the textbook implementation.

This is not your daily PRNG, as it is memory hungry. The internal state requires N \* `sha256_context_t`, where N is typically >= 16, or >= 2.5KB RAM. I wasn't sure to PR this, but after the summit I thought: why not? After all, it fits perfectly in the discussion of #5659. 

It is a WIP, because the reference implementation needs AES-256 (#4798, #5429). Now it uses AES-128, which works fine. Furthermore, initialization of the PRNG requires a seed >> 32 bits, so I cheat a bit. #4294 would help.

I'm planning to do some offline testing, eventually.

### CODE REVIEW

File: Makefile.dep [hunk 1]
```diff
@@ -564,6 +564,14 @@ ifneq (,$(filter random,$(USEMODULE)))
         USEMODULE += prng_tinymt32
     endif
 
+    ifneq (,$(filter prng_fortuna,$(USEMODULE)))
+        USEMODULE += fortuna
+        USEMODULE += hashes
+        USEMODULE += crypto
+        USEMODULE += xtimer
+        CFLAGS += -DCRYPTO_AES
```
Comment (BytesGalore) [inline Makefile.dep]: I wonder if it would be more intuitive to set the CFLAG in `sys/random/Makefile`,
since here we only handle dependencies.
Beside, please remove the extra whitespace behind the flag.

File: sys/random/fortuna/fortuna.c [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    if (state->pools[0].len >= 64 && diff > FORTUNA_RESEED_TIMEOUT) {
+        state->last_reseed = xtimer_now64();
+#else
+    if (state->pools[0].len >= 64) {
+#endif
+        state->reseeds++;
+        size_t len = 0;
+
+        for (size_t i = 0; i < FORTUNA_POOLS; i++) {
+            if (state->reseeds | (1 << i)) {
```
Comment (BytesGalore) [inline sys/random/fortuna/fortuna.c]: isn't this always true?

File: tests/prng_tools/main.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ */
+static void print_distribution(uint32_t* values, uint8_t bins)
+{
+    printf("For %d bit samples:\n", bins);
+
+    /* sum the total number of ones */
+    uint64_t total = 0;
+
+    for (int i = 0; i < bins; i++) {
+        total = total + values[i];
```
Comment (BytesGalore) [inline tests/prng_tools/main.c]: how about `+=`

File: tests/prng_tools/main.c [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /* if total is zero, it would yield a division by zero */
```
Comment (BytesGalore) [inline tests/prng_tools/main.c]: then I propose to `assert` it here

**Comment by BytesGalore** [comment]
nice one, seems to distribute the bits quite equally :+1:

File: Makefile.dep (hunk 1 above)
Comment (basilfx) [inline Makefile.dep]: I think the `CFLAGS` is ugly here, but it's the only way to tell the `crypto` module to include AES. It's WIP after all :-)

File: sys/random/fortuna/fortuna.c (hunk 2 above)
Comment (basilfx) [inline sys/random/fortuna/fortuna.c]: Nope, `state->reseeds` increments over time: this check tries to balance the use of each pool IIRC.

File: tests/prng_tools/main.c (hunk 4 above)
Comment (basilfx) [inline tests/prng_tools/main.c]: IMHO `assert` would be a program error, where this case is a likely scenario (for a bad RNG).

**Comment by basilfx** [comment]
Fixed few comments + rebase. Still WIP (but not abandoned).

**Comment by basilfx** [comment]
For the record, this 33c3 talk may be of interest for this PR: [Wheel of Fortune - Analyzing Embedded OS Random Number Generators](https://media.ccc.de/v/33c3-7949-wheel_of_fortune).

**Comment by miri64** [comment]
For now: needs rebase.

**Comment by basilfx** [comment]
Rebased.

**Comment by basilfx** [comment]
I have updated the PR and I have re-checked my implementation with the [textbook](https://www.schneier.com/academic/paperfiles/fortuna.pdf) and couldn't spot any mistakes.

I have exposed more configuration options as defines and set them to recommended defaults (which is not memory friendly, but can be overridden).

I have removed the WIP label since I think this implementation is ready. It still depend on AES-256, but it works fine with AES-128 at the expense of some security. I don't fully understand the mathematical details, but having a 256-bit key instead of 128-bit reduces the chance of selecting a permutation of input -> outputs where collisions occur (please correct me!). That said, if AES-256 gets contributed, then it is a matter of changing one define.

**Comment by basilfx** [comment]
I decided to take out the testing tools and move it to #8182, so back to WIP and waiting for pr.

File: Makefile.dep (hunk 1 above)
Comment (PeterKietzmann) [reply Makefile.dep]: How about sys/Makefile.dep?

**Comment by basilfx** [comment]
Still working on this one. I've removed `rng_tools`, which was introduced using #8182. I still have to fix one remaining issue to seed the generator on start.

**Comment by PeterKietzmann** [comment]
@basilfx thanks for the heads up. What exactly is the issue with seeding? Seeding this kind of generators is still a somehow open issue that also applies to other RNGs. Can we address it in a separate issue/PR?

**Comment by basilfx** [comment]
I'm currently stuck on [this](https://github.com/basilfx/RIOT/blob/d80210987953a06c801514a96b4abaa8b1b2c330/sys/random/fortuna/fortuna.c#L139) piece of code.

The problem is that Fortuna expects some random events to initialize the pools. And that is something we don't do (yet). So I'm looking into other Fortuna implementations how they handle that, without deviating too much from the original implementation.

Last November was the last time I looked into this, so I have to get things sharp again :-)

**Comment by basilfx** [comment]
@PeterKietzmann I have posted a question [here](https://crypto.stackexchange.com/questions/56390/seeding-the-fortuna-prng-for-the-first-time) for the remaining issue I have, regarding 'first boot'.

**Comment by basilfx** [comment]
I have finished work on this PR and it is ready for review. Since a lot has changed since the initial feedback, I have squashed all commits that were not applicable anymore.

This PRNG will still be memory hungry, and the number of pools matter. I default to 32 (paper suggested), but you will have to lower to ~10 for a MCU with at least 32 KiB.

To test this PR, use `tests/rng` and compile it with something like this: `USEMODULE=prng_fortuna [CFLAGS="-DFORTUNA_POOLS=10"] make` (`CFLAGS` not needed for native).

**Comment by smlng** [comment]
building this for SAMR21-XPRO fails: 
```
BOARD=samr21-xpro USEMODULE=prng_fortuna make all
Building application "tests_rng" for "samr21-xpro" with MCU "samd21".

[...]

/RIOT/tests/rng/bin/samr21-xpro/random.a(fortuna.o): In function `random_bytes':
/RIOT/sys/random/fortuna.c:102: multiple definition of `random_bytes'
/RIOT/tests/rng/bin/samr21-xpro/random.a(random.o):/RIOT/sys/random/random.c:44: first defined here
collect2: error: ld returned 1 exit status
make: *** [link] Error 1
```

also static-tests by travis complain about licence header stuff.

File: Makefile.dep [hunk 5]
```diff
@@ -5,7 +5,8 @@ OLD_USEPKG := $(sort $(USEPKG))
 # include board dependencies
 -include $(RIOTBOARD)/$(BOARD)/Makefile.dep
 
-# pull dependencies from drivers
+# pull dependencies from sys and drivers
+include $(RIOTBASE)/sys/Makefile.dep
```
Comment (smlng) [inline Makefile.dep]: not sure about introducing this here too. Though it is consistent with `/drivers`, but should still go in a separate PR and then adapt all sys modules deps, too.

File: Makefile.dep [hunk 6]
```diff
@@ -607,6 +608,13 @@ ifneq (,$(filter random,$(USEMODULE)))
     USEMODULE += prng_tinymt32
   endif
 
+  ifneq (,$(filter prng_fortuna,$(USEMODULE)))
```
Comment (smlng) [inline Makefile.dep]: btw. with the above applied, this would go into `sys/Makefile.dep` right?

File: sys/random/fortuna/fortuna.h [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+ */
+typedef uint32_t fortuna_seed_t[FORTUNA_SEED_SIZE];
+
+/**
+ * @brief   Initialize the Fortuna PRNG state.
+ *
+ * It is possible to use this method to clear an existing state.
+ *
+ * @param[in] state         PRNG state
+ * @return                  zero on succesful initialization
```
Comment (smlng) [inline sys/random/fortuna/fortuna.h]: here and below, add empty line between param and return

File: sys/random/fortuna/fortuna.h [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+ *
+ * @param[in] state         PRNG state
+ * @param[out] out          pointer to buffer
+ * @param[in] bytes         number of bytes to write in buffer
+ * @return                  zero on success
+ * @return                  -1 on reading more that FORTUNA_RESEED_LIMIT bytes
+ * @return                  -2 on reading from unseeded PRNG
+ * @returned                -3 on other error
+ */
+int fortuna_random_data(fortuna_state_t *state, uint8_t *out, size_t bytes);
```
Comment (smlng) [inline sys/random/fortuna/fortuna.h]: consider making `state` const if it is input param only.

File: sys/random/fortuna/fortuna.h [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+ * @param[in] state         PRNG state
+ * @param[in] data          pointer to entropy source
+ * @param[in] length        length of entropy source
+ * @param[in] source        source identifier (each source has its own ID)
+ * @param[in] pool          pool number to add entropy to
+ * @return                  zero on success
+ * @return                  -1 on zero bytes or more than 32 bytes
+ * @return                  -2 on invalid pool number
+ */
+int fortuna_add_random_event(fortuna_state_t *state, uint8_t *data,
```
Comment (smlng) [inline sys/random/fortuna/fortuna.h]: same here (and below) consider making input params (pointers) const if possible.

File: Makefile.dep (hunk 5 above)
Comment (basilfx) [reply Makefile.dep]: I'm not sure what I should adapt. I could move it into a separate PR, but the only entry in `sys/Makefile.dep` is now Fortuna PRNG related. That would not make sense.

File: Makefile.dep (hunk 6 above)
Comment (basilfx) [reply Makefile.dep]: I think so. Not sure, because I basically duplicated `prng_tinymt32`.

File: sys/random/fortuna/fortuna.h (hunk 7 above)
Comment (basilfx) [reply sys/random/fortuna/fortuna.h]: Done.

File: sys/random/fortuna/fortuna.h (hunk 8 above)
Comment (basilfx) [reply sys/random/fortuna/fortuna.h]: Tried, but doesn't work since state is modified. I therefore changed it to `@param[inout]`.

File: sys/random/fortuna/fortuna.h (hunk 9 above)
Comment (basilfx) [reply sys/random/fortuna/fortuna.h]: Same.

File: Makefile.dep (hunk 5 above)
Comment (basilfx) [reply Makefile.dep]: I've moved it to `sys/Makefile.dep` because of https://github.com/RIOT-OS/RIOT/pull/5675#discussion_r157827407.

**Comment by basilfx** [comment]
@smlng Thanks for the review. I have added additional commits to fix the issues.

File: Makefile.dep (hunk 5 above)
Comment (smlng) [reply Makefile.dep]: okay leave as is, I was just think that other sys-modules could be adapted, i.e., move deps form `RIOT/Makefile.dep` to `RIOT/sys/Makefile.dep` - but obviously not in this PR 😄

File: Makefile.dep (hunk 6 above)
Comment (smlng) [reply Makefile.dep]: see comment above, leave as is for now and we adapt all sys-modules in a separate cleanup-PR.

File: sys/random/fortuna/fortuna.h (hunk 8 above)
Comment (smlng) [reply sys/random/fortuna/fortuna.h]: good

File: sys/random/fortuna.c [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * @brief Wrapper for fortuna_random_data that supports reading more than
+ *        FORTUNA_RESEED_LIMIT and also checks the result.
+ */
+static void _read(uint8_t *out, size_t bytes)
+{
+    do {
+        /* read at most chunk bytes to not exhaust the state at once */
+        size_t chunk = bytes < FORTUNA_RESEED_LIMIT ?
```
Comment (smlng) [inline sys/random/fortuna.c]: put condition in parentheses, maybe?

File: sys/random/fortuna.c [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+                       bytes : FORTUNA_RESEED_LIMIT;
+
+        int res = fortuna_random_data(&fortuna_state, out, bytes);
```
Comment (smlng) [inline sys/random/fortuna.c]: shouldn't this use `chunk` instead of `bytes`, i.e., `fortuna_random_data(&fortuna_state, out, chunk)`?

Comment (basilfx) [reply sys/random/fortuna.c]: Yes, good catch!

**Comment by basilfx** [comment]
@smlng Rebased and fixed the two comments.

**Comment by smlng** [comment]
please squash, and the we give murdock a try

**Comment by basilfx** [comment]
Murdock is happy.
